From 35b4a8f0464fa7a2f01d503cfa637bdcd7f69b3e Mon Sep 17 00:00:00 2001
From: Kazuhiko SHIOZAKI <kazuhiko@nexedi.com>
Date: Tue, 8 Apr 2025 11:31:55 +0300
Subject: [PATCH] repozo: introduce --keep-only-latest-index option in backup
 to make rsync'ish backup more efficient.

If this option is enabled:
* keep the latest index only.
* always use the fixed filename for the latest index and make a symlink
  to it having the original datetime based filename.
---
 src/ZODB/scripts/repozo.py            | 36 ++++++++++++++++++++++++++-
 src/ZODB/scripts/tests/test_repozo.py |  3 +++
 2 files changed, 38 insertions(+), 1 deletion(-)

diff --git a/src/ZODB/scripts/repozo.py b/src/ZODB/scripts/repozo.py
index 8cc5dc79..c313284b 100755
--- a/src/ZODB/scripts/repozo.py
+++ b/src/ZODB/scripts/repozo.py
@@ -58,6 +58,12 @@ Options for -B/--backup:
         backup files (and associated metadata files) from the repository
         directory.
 
+    -K / --keep-only-latest-index
+        If a full backup is created, keep the latest index file only as the
+        fixed filename and make a symlink to it having the original datetime
+        based filename. Enabling this will reduce the date transfer using
+        rsync'ish backup.
+
 Options for -R/--recover:
     -D str
     --date=str
@@ -101,6 +107,7 @@ from __future__ import print_function
 
 import errno
 import getopt
+import glob
 import gzip
 import os
 import re
@@ -178,7 +185,7 @@ def error(msg, *args):
 def parseargs(argv):
     global VERBOSE
     try:
-        opts, args = getopt.getopt(argv, 'BRVvhr:f:FQzkD:o:w',
+        opts, args = getopt.getopt(argv, 'BRVvhr:f:FQzkD:o:wK',
                                    ['backup',
                                     'recover',
                                     'verify',
@@ -193,6 +200,7 @@ def parseargs(argv):
                                     'date=',
                                     'output=',
                                     'with-verify',
+                                    'keep-only-latest-index',
                                     ])
     except getopt.error as msg:
         usage(1, msg)
@@ -208,6 +216,7 @@ def parseargs(argv):
         gzip = False        # -z flag state
         killold = False     # -k flag state
         withverify = False  # -w flag state
+        keep_only_latest_index = False  # -K flag state
 
     options = Options()
 
@@ -246,6 +255,8 @@ def parseargs(argv):
             options.killold = True
         elif opt in ('-w', '--with-verify'):
             options.withverify = True
+        elif opt in ('-K', '--keep-only-latest-index'):
+            options.keep_only_latest_index = True
         else:
             assert False, (opt, arg)
 
@@ -280,6 +291,9 @@ def parseargs(argv):
         if options.full and options.quick:
             log('--quick option is ignored if --full option is used')
             options.quick = None
+        if options.keep_only_latest_index:
+            log('--keep-only-latest-index option is ignored in recover mode')
+            options.keep_only_latest_index = False
     else:
         assert options.mode == VERIFY
         if options.date is not None:
@@ -303,6 +317,9 @@ def parseargs(argv):
         if options.withverify is not None:
             log('--with-verify option is ignored in verify mode')
             options.withverify = None
+        if options.keep_only_latest_index:
+            log('--keep-only-latest-index option is ignored in recover mode')
+            options.keep_only_latest_index = False
     return options
 
 
@@ -600,6 +617,8 @@ def do_full_backup(options):
     log('writing index')
     fs._index.save(pos, index_file)
     fs.close()
+    if options.keep_only_latest_index:
+        keep_only_latest_index(options)
     log('writing full backup: %s bytes to %s', pos, dest)
     sum = copyfile(options, dest, 0, pos)
     # Write the data file for this full backup
@@ -632,6 +651,8 @@ def do_incremental_backup(options, reposz, repofiles):
                               gen_filename(options, '.index', tnow))
     fs._index.save(pos, index_file)
     fs.close()
+    if options.keep_only_latest_index:
+        keep_only_latest_index(options)
     log('writing incremental: %s bytes to %s',  pos-reposz, dest)
     sum = copyfile(options, dest, reposz, pos - reposz)
     # The first file in repofiles points to the last full backup.  Use this to
@@ -647,6 +668,19 @@ def do_incremental_backup(options, reposz, repofiles):
     fp.close()
 
 
+def keep_only_latest_index(options):
+    index_file_list = glob.glob(os.path.join(options.repository, '*.index'))
+    index_file_list.sort()
+    for index_file in index_file_list[:-1]:
+        os.remove(index_file)
+    if not os.path.islink(index_file_list[-1]):
+        latest_index_file = os.path.join(options.repository, 'index.latest')
+        if os.path.isfile(latest_index_file):
+            os.remove(latest_index_file)
+        os.rename(index_file_list[-1], latest_index_file)
+        os.symlink(os.path.basename(latest_index_file), index_file_list[-1])
+
+
 def do_backup(options):
     repofiles = find_files(options)
     # See if we need to do a full backup
diff --git a/src/ZODB/scripts/tests/test_repozo.py b/src/ZODB/scripts/tests/test_repozo.py
index 0a9b48e8..0cb084a9 100644
--- a/src/ZODB/scripts/tests/test_repozo.py
+++ b/src/ZODB/scripts/tests/test_repozo.py
@@ -348,6 +348,7 @@ class OptionsTestBase(object):
         class Options(object):
             repository = self._repository_directory
             date = None
+            keep_only_latest_index = False
 
             def __init__(self, **kw):
                 self.__dict__.update(kw)
@@ -791,6 +792,7 @@ class Test_do_incremental_backup(OptionsTestBase, unittest.TestCase):
                                     killold=False,
                                     test_now=(2010, 5, 14, 10, 51, 22),
                                     date=None,
+                                    keep_only_latest_index=True,
                                     )
         fullfile = os.path.join(self._repository_directory,
                                 '2010-05-14-00-00-00.fs')
@@ -828,6 +830,7 @@ class Test_do_incremental_backup(OptionsTestBase, unittest.TestCase):
                                     killold=False,
                                     test_now=(2010, 5, 14, 10, 51, 22),
                                     date=None,
+                                    keep_only_latest_index=True,
                                     )
         fullfile = os.path.join(self._repository_directory,
                                 '2010-05-14-00-00-00.fs')
-- 
2.48.0

